contracts/sui/xcall/sources/main.move

```diff
@@ -0,0 +1,547 @@
+        let from = message_request::from(proxy_request);
+        let to = message_request::to(proxy_request);
+        let sn = message_request::sn(proxy_request);
+        let msg_type = message_request::msg_type(proxy_request);
+        let msg_data_hash=message_request::data(proxy_request);
+        let protocols = message_request::protocols(proxy_request);
+
+        let data_hash = hash::keccak256(&data);
+        assert!(msg_data_hash == data_hash, EDataMismatch);
+        assert!(to==utils::id_to_hex_string(&xcall_state::get_id_cap_id(cap)),EInvalidAccess);
+        if(msg_type==call_message_rollback::msg_type()){
+            xcall_state::set_reply_state(self, *proxy_request);
+        };
+        let ticket=execute_ticket::new(
+            xcall_state::get_id_cap_id(cap),
+            request_id,
+            from,
+            protocols,
+            data,
+        );   
+        ticket
+    }
+
+
+    public fun execute_call_result(self:&mut Storage,ticket:ExecuteTicket,success:bool,mut fee:Coin<SUI>,ctx:&mut TxContext){
+        self.enforce_version(CURRENT_VERSION);
+        let request_id=ticket.request_id();
+        let proxy_request = xcall_state::get_proxy_request(self, request_id);
+        let msg_type = message_request::msg_type(proxy_request);
+        let sn = message_request::sn(proxy_request);
+        let from=message_request::from(proxy_request);
+        let net_to=network_address::net_id(&from);
+        let protocols=message_request::protocols(proxy_request);
+
+        if(msg_type==persistent_message::msg_type() && !success){
+            abort 0x01
+        };
+        xcall_state::remove_proxy_request(self, request_id);
+        xcall_state::remove_reply_state(self);
+        let mut message = vector::empty<u8>();
+        let code= if(success){1}else{0};
+        let err_msg = if(success){string::utf8(b"success")}else{string::utf8(b"unknown error")};
+        let cs_message_result = if(msg_type==call_message_rollback::msg_type()){
+            let callReply = xcall_state::get_call_reply(self);
+            if(vector::length(&callReply) > 0 && code == 1){
+                message = callReply;
+                xcall_state::remove_call_reply(self);
+            };
+            message_result::create(sn, code, message)
+
+        }else {
+            message_result::create(sn, code, message)
+        };
+        execute_ticket::consume(ticket);
+        if(msg_type==call_message_rollback::msg_type()){
+          let cs_message=cs_message::from_message_result(cs_message_result);
+           fee = connection_send_message(self,fee,protocols,net_to,cs_message,sn,true,ctx);
+        };
+
+        transfer::public_transfer(fee, ctx.sender());
+        event::emit(CallExecuted{req_id:request_id, code: code, err_msg: err_msg});
+       
+    }
+
+    public fun execute_rollback(self:&mut Storage,cap:&IDCap, sn:u128,ctx: &mut TxContext):RollbackTicket{
```

Comment:
Can a dapp here re execute this a claim multiple tickets for the same message

Reply:
ya it can take multiple ticket and call rollback execute multiple times with error but for success it can call only once.